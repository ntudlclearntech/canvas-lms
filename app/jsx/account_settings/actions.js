/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

function pluralize(word) {
  if (word[word.length - 1] !== 's') {
    return `${word}s`
  }
  return word
}

export const SET_CSP_ENABLED = 'SET_CSP_ENABLED'
export const SET_CSP_ENABLED_OPTIMISTIC = 'SET_CSP_ENABLED_OPTIMISTIC'

export function setCspEnabledAction(value, opts = {}) {
  const type = opts.optimistic ? SET_CSP_ENABLED_OPTIMISTIC : SET_CSP_ENABLED
  if (typeof value !== 'boolean') {
    return {
      type,
      payload: new Error('Can only set to Boolean values'),
      error: true
    }
  }
  return {
    type,
    payload: value
  }
}

export function setCspEnabled(context, contextId, value) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) => {
    dispatch(setCspEnabledAction(value, {optimistic: true}))
    return axios
      .put(`/api/v1/${context}/${contextId}/csp_settings`, {
        status: value ? 'enabled' : 'disabled'
      })
      .then(response => {
        dispatch(setCspEnabledAction(response.data.enabled))
      })
  }
}

export function getCspEnabled(context, contextId) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) =>
    axios.get(`/api/v1/${context}/${contextId}/csp_settings`).then(response => {
      dispatch(setCspEnabledAction(response.data.enabled))
    })
}

export const SET_CSP_INHERITED = 'SET_CSP_INHERITED'
export const SET_CSP_INHERITED_OPTIMISTIC = 'SET_CSP_INHERITED_OPTIMISTIC'

export function setCspInheritedAction(value, opts = {}) {
  const type = opts.optimistic ? SET_CSP_INHERITED_OPTIMISTIC : SET_CSP_INHERITED
  if (typeof value !== 'boolean') {
    return {
      type,
      payload: new Error('Can only set to Boolean values'),
      error: true
    }
  }
  return {
    type,
    payload: value
  }
}

export function setCspInherited(context, contextId, value) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) => {
    dispatch(setCspInheritedAction(value, {optimistic: true}))
    if (value) {
      return axios
        .put(`/api/v1/${context}/${contextId}/csp_settings`, {
          status: 'inherited'
        })
        .then(response => {
          dispatch(setCspInheritedAction(response.data.inherited))
        })
    }
  }
}

export function getCspInherited(context, contextId) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) =>
    axios.get(`/api/v1/${context}/${contextId}/csp_settings`).then(response => {
      dispatch(setCspInheritedAction(response.data.inherited))
    })
}

export const ADD_DOMAIN = 'ADD_DOMAIN'
export const ADD_DOMAIN_BULK = 'ADD_DOMAIN_BULK'
export const ADD_DOMAIN_OPTIMISTIC = 'ADD_DOMAIN_OPTIMISTIC'
const DOMAIN_MAP_KEYS = ['account', 'tools', 'effective']

export function addDomainAction(domain, domainType, opts = {}) {
  const type = opts.optimistic ? ADD_DOMAIN_OPTIMISTIC : ADD_DOMAIN
  if (typeof domain !== 'string') {
    return {
      type,
      payload: new Error('Can only set to String values'),
      error: true
    }
  }
  if (!DOMAIN_MAP_KEYS.includes(domainType)) {
    return {
      type,
      payload: new Error('domainType is invalid'),
      error: true
    }
  }
  return {
    type,
    payload: {[domainType]: domain}
  }
}

export function addDomainBulkAction(domainsMap) {
  if (Object.keys(domainsMap).some(d => !DOMAIN_MAP_KEYS.includes(d))) {
    return {
      type: ADD_DOMAIN_BULK,
      payload: new Error('Invalid domain type key provided in domainsMap'),
      error: true
    }
  }
  return {
    type: ADD_DOMAIN_BULK,
    payload: domainsMap
  }
}

export function addDomain(context, contextId, domain) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) => {
    dispatch(addDomainAction(domain, 'account', {optimistic: true}))
    return axios
      .post(`/api/v1/${context}/${contextId}/csp_settings/domains`, {
        domain
      })
      .then(() => {
        // This isn't really necessary but since the whitelist is unique,
        // it doesn't hurt.
        dispatch(addDomainAction(domain, 'account'))
      })
  }
}

export function getCurrentWhitelist(context, contextId) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) =>
    axios.get(`/api/v1/${context}/${contextId}/csp_settings`).then(response => {
      const addDomainMap = {
        effective: response.data.effective_whitelist || [],
        account: response.data.current_account_whitelist || [],
        tools: response.data.tools_whitelist || {}
      }
      dispatch(addDomainBulkAction(addDomainMap))
    })
}

export const REMOVE_DOMAIN = 'REMOVE_DOMAIN'
export const REMOVE_DOMAIN_OPTIMISTIC = 'REMOVE_DOMAIN_OPTIMISTIC'

export function removeDomainAction(domain, opts = {}) {
  const type = opts.optimistic ? REMOVE_DOMAIN_OPTIMISTIC : REMOVE_DOMAIN
  if (typeof domain !== 'string') {
    return {
      type,
      payload: new Error('Domain can only set to String values'),
      error: true
    }
  }
  return {
    type,
    payload: domain
  }
}

export function removeDomain(context, contextId, domain) {
  context = pluralize(context)
  return (dispatch, getState, {axios}) => {
    dispatch(removeDomainAction(domain, {optimistic: true}))
    return axios
      .delete(`/api/v1/${context}/${contextId}/csp_settings/domains?domain=${domain}`)
      .then(() => {
        // This isn't really necessary but doesn't hurt
        dispatch(removeDomainAction(domain))
      })
  }
}
