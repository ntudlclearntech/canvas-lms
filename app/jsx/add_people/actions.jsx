define([
  'redux-actions',
  './api_client'
], ({ createActions }, api) => {
  const actionDefs = [
    'SET_INPUT_PARAMS',

    'VALIDATE_USERS_START',     // validate users api lifecycle
    'VALIDATE_USERS_SUCCESS',
    'VALIDATE_USERS_ERROR',

    'CREATE_USERS_START',       // invite users api lifecycle
    'CREATE_USERS_SUCCESS',
    'CREATE_USERS_ERROR',

    'ENROLL_USERS_START',       // enrols users api lifecycle
    'ENROLL_USERS_SUCCESS',
    'ENROLL_USERS_ERROR',

    'CHOOSE_DUPLICATE',           // choose from a set of duplicates
    'SKIP_DUPLICATE',             // skip this set of duplicates
    'ENQUEUE_NEW_FOR_DUPLICATE',  // prepare to create a new user in lieu of one of the duplicates
    'ENQUEUE_NEW_FOR_MISSING',    // prepare to create a new user for one of the missing users

    'ENQUEUE_USERS_TO_BE_ENROLLED', // prepare to enroll validated users

    'RESET'   // reset([array of state subsections to reset]) undefined or empty = reset everything
  ];

  const actionTypes = actionDefs.reduce((types, action) => {
    types[action] = action;   // eslint-disable-line no-param-reassign
    return types;
  }, {});

  const actions = createActions(...actionDefs);

  actions.validateUsers = () => (dispatch, getState) => {
    dispatch(actions.validateUsersStart());
    const state = getState();
    const courseId = state.courseParams.courseId;
    const users = state.inputParams.nameList;
    const searchType = state.inputParams.searchType;
    api.validateUsers({ courseId, users, searchType })
      .then((res) => {
        dispatch(actions.validateUsersSuccess(res.data));
        // if all the users were found, then we can jump right to enrolling
        if (res.data.duplicates.length === 0 && res.data.missing.length === 0) {
          const st = getState();
          dispatch(actions.enqueueUsersToBeEnrolled(st.userValidationResult.validUsers));
        }
      })
      .catch((err) => {
        dispatch(actions.validateUsersError(err));
      });
  };


  actions.resolveValidationIssues = () => (dispatch, getState) => {
    dispatch(actions.createUsersStart());
    const state = getState();
    const courseId = state.courseParams.courseId;

    // the list of users to be enrolled
    let usersToBeEnrolled = state.userValidationResult.validUsers.slice();
    // and the list of users to be created
    const usersToBeCreated = [];

    Object.keys(state.userValidationResult.duplicates).forEach((addr) => {
      const dupeSet = state.userValidationResult.duplicates[addr];
      if (dupeSet.createNew && dupeSet.newUserInfo.name && dupeSet.newUserInfo.email) {
        usersToBeCreated.push(dupeSet.newUserInfo);
      } else if (dupeSet.selectedUserId >= 0) {
        const selectedUser = dupeSet.userList.find(u => u.user_id === dupeSet.selectedUserId);
        usersToBeEnrolled.push(selectedUser);
      }
    });
    Object.keys(state.userValidationResult.missing).forEach((addr) => {
      const missing = state.userValidationResult.missing[addr];
      if (missing.createNew && missing.newUserInfo.name && missing.newUserInfo.email) {
        usersToBeCreated.push(missing.newUserInfo);
      }
    });

    api.createUsers({ courseId, users: usersToBeCreated })
      .then((res) => {
        dispatch(actions.createUsersSuccess(res.data));
        // merge in the newly created users
        usersToBeEnrolled = usersToBeEnrolled.concat(res.data.invited_users.map((u) => {
          // adjust shape of users we just invited to match the existing users
          const user = { ...u }
          user.user_name = u.name;
          user.address = u.email;
          return user;
        }));
        dispatch(actions.enqueueUsersToBeEnrolled(usersToBeEnrolled));
      })
      .catch(err => dispatch(actions.createUsersError(err)));
  };

  actions.enrollUsers = () => (dispatch, getState) => {
    dispatch(actions.enrollUsersStart());
    const state = getState();
    const courseId = state.courseParams.courseId;
    const users = state.usersToBeEnrolled.map(u => u.user_id);
    const role = state.inputParams.role
          || (state.courseParams.roles && state.courseParams.roles.length && state.courseParams.roles[0].id)
          || '';
    const section = state.inputParams.section
          || (state.courseParams.sections && state.courseParams.sections.length && state.courseParams.sections[0].id)
          || '';
    const limitPrivilege = state.inputParams.limitPrivilege || false;
    api.enrollUsers({ courseId, users, role, section, limitPrivilege })
      .then(res => dispatch(actions.enrollUsersSuccess(res.data)))
      .catch(err => dispatch(actions.enrollUsersError(err)));
  };
  return { actions, actionTypes };
});
