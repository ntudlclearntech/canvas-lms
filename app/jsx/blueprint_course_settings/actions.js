import { createActions } from 'redux-actions'
import api from './apiClient'

const types = [
  'LOAD_COURSES_START', 'LOAD_COURSES_SUCCESS', 'LOAD_COURSES_FAIL',
  'LOAD_ASSOCIATIONS_START', 'LOAD_ASSOCIATIONS_SUCCESS', 'LOAD_ASSOCIATIONS_FAIL',
  'SAVE_ASSOCIATIONS_START', 'SAVE_ASSOCIATIONS_SUCCESS', 'SAVE_ASSOCIATIONS_FAIL',
  'CHECK_MIGRATION_START', 'CHECK_MIGRATION_SUCCESS', 'CHECK_MIGRATION_FAIL',
  'BEGIN_MIGRATION_START', 'BEGIN_MIGRATION_SUCCESS', 'BEGIN_MIGRATION_FAIL',
  'ADD_COURSE_ASSOCIATIONS', 'UNDO_ADD_COURSE_ASSOCIATIONS',
  'REMOVE_COURSE_ASSOCIATIONS', 'UNDO_REMOVE_COURSE_ASSOCIATIONS',
  'CLEAR_ASSOCIATIONS',
]
const actions = createActions(...types)

actions.checkMigration = () => (dispatch, getState) => {
  dispatch(actions.checkMigrationStart())
  api.checkMigration(getState())
    .then(res => dispatch(actions.checkMigrationSuccess(res.data)))
    .catch(err => dispatch(actions.checkMigrationFail(err)))
}

actions.beginMigration = () => (dispatch, getState) => {
  dispatch(actions.beginMigrationStart())
  api.beginMigration(getState())
    .then(res => dispatch(actions.beginMigrationSuccess(res.data)))
    .catch(err => dispatch(actions.beginMigrationFail(err)))
}

actions.addAssociations = associations => (dispatch, getState) => {
  const state = getState()
  const existing = state.existingAssociations
  const toAdd = []
  const toUndo = []

  associations.forEach((courseId) => {
    if (existing.find(course => course.id === courseId)) {
      toUndo.push(courseId)
    } else {
      toAdd.push(courseId)
    }
  })

  if (toAdd.length) {
    const courses = state.courses.concat(state.existingAssociations)
    dispatch(actions.addCourseAssociations(courses.filter(c => toAdd.includes(c.id))))
  }

  if (toUndo.length) {
    dispatch(actions.undoRemoveCourseAssociations(toUndo))
  }
}

actions.removeAssociations = associations => (dispatch, getState) => {
  const existing = getState().existingAssociations
  const toRm = []
  const toUndo = []

  associations.forEach((courseId) => {
    if (existing.find(course => course.id === courseId)) {
      toRm.push(courseId)
    } else {
      toUndo.push(courseId)
    }
  })

  if (toRm.length) {
    dispatch(actions.removeCourseAssociations(toRm))
  }

  if (toUndo.length) {
    dispatch(actions.undoAddCourseAssociations(toUndo))
  }
}

actions.loadCourses = filters => (dispatch, getState) => {
  dispatch(actions.loadCoursesStart())
  api.getCourses(getState(), filters)
    .then(res => dispatch(actions.loadCoursesSuccess(res.data)))
    .catch(err => dispatch(actions.loadCoursesFail(err)))
}

actions.loadAssociations = () => (dispatch, getState) => {
  dispatch(actions.loadAssociationsStart())
  api.getAssociations(getState())
    .then((res) => {
      const data = res.data.map(course =>
        Object.assign({}, course, {
          term: {
            id: '0',
            name: course.term_name,
          },
          term_name: undefined,
        }))
      dispatch(actions.loadAssociationsSuccess(data))
    })
    .catch(err => dispatch(actions.loadAssociationsFail(err)))
}

actions.saveAssociations = () => (dispatch, getState) => {
  dispatch(actions.saveAssociationsStart())
  const state = getState()
  api.saveAssociations(state)
    .then(() => dispatch(actions.saveAssociationsSuccess({ added: state.addedAssociations, removed: state.removedAssociations })))
    .catch(err => dispatch(actions.saveAssociationsFail(err)))
}

const actionTypes = types.reduce((typesMap, actionType) =>
  Object.assign(typesMap, { [actionType]: actionType }), {})

export { actionTypes, actions as default }
