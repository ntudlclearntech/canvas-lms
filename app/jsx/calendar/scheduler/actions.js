import { createAction } from 'redux-actions'

  const keys = {
    SET_FIND_APPOINTMENT_MODE: 'SET_FIND_APPOINTMENT_MODE',
    SET_COURSE: 'SET_COURSE'
  }

  const actions = {
    setFindAppointmentMode: createAction(keys.SET_FIND_APPOINTMENT_MODE),
    setCourse: createAction(keys.SET_COURSE)
  };

export default {
    actions,
    keys
  };
