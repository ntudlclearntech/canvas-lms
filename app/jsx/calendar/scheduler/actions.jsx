define([
  'redux-actions',
], (ReduxActions) => {

  const { createAction } = ReduxActions;

  const keys = {
    SET_FIND_APPOINTMENT_MODE: 'SET_FIND_APPOINTMENT_MODE'
  }

  const actions = {
    setFindAppointmentMode: createAction(keys.SET_FIND_APPOINTMENT_MODE)
  };

  return {
    actions,
    keys
  };

});
