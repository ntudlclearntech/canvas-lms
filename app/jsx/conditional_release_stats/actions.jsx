define([
  './cyoe-api',
  './helpers/actions',
], (cyoeClient, { createActions }) => {
  const actionDefs = [
    'SET_SCORING_RANGES',
    'SET_RULE',
    'SET_ENROLLED',
    'SET_ASSIGNMENT',
    'SET_ERRORS',
    'SELECT_RANGE',
    'SELECT_STUDENT',
    'OPEN_SIDEBAR',
    'CLOSE_SIDEBAR',
    'LOAD_START',
    'LOAD_END',
  ]

  const { actions, actionTypes } = createActions(actionDefs)

  actions.loadData = (assignment) => {
    return (dispatch, getState) => {
      dispatch(actions.loadStart())

      cyoeClient.loadInitialData(getState())
        .then(data => {
          dispatch(actions.setScoringRanges(data.ranges))
          dispatch(actions.setRule(data.rule))
          dispatch(actions.setEnrolled(data.enrolled))
          dispatch(actions.loadEnd())
        }, errors => {
          dispatch(actions.setErrors(errors))
          dispatch(actions.loadEnd())
        })
    }
  }

  return { actions, actionTypes }
})
