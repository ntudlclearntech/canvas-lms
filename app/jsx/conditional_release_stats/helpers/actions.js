import _ from 'underscore'

export function typeNameToFuncName (typeName) {
  const parts = typeName.split('_')
  return parts.map((part, i) => {
    if (i !== 0 && part.length) {
      return part.charAt(0).toUpperCase() + part.slice(1).toLowerCase()
    } else {
      return part.toLowerCase()
    }
  }).join('')
}

export function createAction (actionType) {
  const actionCreator = (payload) => ({
    type: actionType,
    payload,
  })

  actionCreator.type = actionType
  actionCreator.toString = () => actionType
  return actionCreator
}

export function createActions (actionDefs) {
  const actionTypes = {}
  const actions = {}

  actionDefs.forEach(actionDef => {
    const action = createAction(actionDef)
    const funcName = typeNameToFuncName(action.type)
    actions[funcName] = action
    actionTypes[action.type] = action.type
  })

  return { actionTypes, actions }
}
