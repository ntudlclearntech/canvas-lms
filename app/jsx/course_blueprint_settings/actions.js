import { createActions } from 'redux-actions'
import api from './apiClient'

const types = [
  'LOAD_COURSES_START', 'LOAD_COURSES_SUCCESS', 'LOAD_COURSES_FAIL',
  'LOAD_ASSOCIATIONS_START', 'LOAD_ASSOCIATIONS_SUCCESS', 'LOAD_ASSOCIATIONS_FAIL',
  'SAVE_ASSOCIATIONS_START', 'SAVE_ASSOCIATIONS_SUCCESS', 'SAVE_ASSOCIATIONS_FAIL',
  'ADD_COURSE_ASSOCIATIONS', 'UNDO_ADD_COURSE_ASSOCIATIONS',
  'REMOVE_COURSE_ASSOCIATIONS', 'UNDO_REMOVE_COURSE_ASSOCIATIONS',
]
const actions = createActions(...types)

actions.cancel = () => (dispatch, getState) => {
  const referrer = document.referrer
  if (referrer.split('/')[2] !== location.host || referrer.includes('/login')) {
    location.replace(`/courses/${getState().course.id}`)
  } else {
    location.replace(referrer)
  }
}

actions.addAssociations = associations => (dispatch, getState) => {
  const state = getState()
  const existing = state.existingAssociations
  const toAdd = []
  const toUndo = []

  associations.forEach((courseId) => {
    if (existing.find(course => course.id === courseId)) {
      toUndo.push(courseId)
    } else {
      toAdd.push(courseId)
    }
  })

  if (toAdd.length) {
    const courses = state.courses.concat(state.existingAssociations)
    dispatch(actions.addCourseAssociations(courses.filter(c => toAdd.includes(c.id))))
  }

  if (toUndo.length) {
    dispatch(actions.undoRemoveCourseAssociations(toUndo))
  }
}

actions.removeAssociations = associations => (dispatch, getState) => {
  const existing = getState().existingAssociations
  const toRm = []
  const toUndo = []

  associations.forEach((courseId) => {
    if (existing.find(course => course.id === courseId)) {
      toRm.push(courseId)
    } else {
      toUndo.push(courseId)
    }
  })

  if (toRm.length) {
    dispatch(actions.removeCourseAssociations(toRm))
  }

  if (toUndo.length) {
    dispatch(actions.undoAddCourseAssociations(toUndo))
  }
}

actions.loadCourses = filters => (dispatch, getState) => {
  dispatch(actions.loadCoursesStart())
  api.getCourses(getState(), filters)
    .then(res => dispatch(actions.loadCoursesSuccess(res.data)))
    .catch(err => dispatch(actions.loadCoursesFail(err)))
}

actions.loadAssociations = () => (dispatch, getState) => {
  dispatch(actions.loadAssociationsStart())
  api.getAssociations(getState())
    .then((res) => {
      const data = res.data.map(course =>
        Object.assign({}, course, {
          term: {
            id: '0',
            name: course.term_name,
          },
          term_name: undefined,
        }))
      dispatch(actions.loadAssociationsSuccess(data))
    })
    .catch(err => dispatch(actions.loadAssociationsFail(err)))
}

actions.saveAssociations = () => (dispatch, getState) => {
  dispatch(actions.saveAssociationsStart())
  const state = getState()
  api.saveAssociations(state)
    .then(() => dispatch(actions.saveAssociationsSuccess({ added: state.addedAssociations, removed: state.removedAssociations })))
    .catch(err => dispatch(actions.saveAssociationsFail(err)))
}

const actionTypes = types.reduce((typesMap, actionType) =>
  Object.assign(typesMap, { [actionType]: actionType }), {})

export { actionTypes, actions as default }
