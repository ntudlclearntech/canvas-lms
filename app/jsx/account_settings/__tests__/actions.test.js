/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import * as Actions from '../actions'

describe('setCspEnabledAction', () => {
  it('creates a SET_CSP_ENABLED action when passed a boolean value', () => {
    expect(Actions.setCspEnabledAction(true)).toMatchSnapshot()
  })
  it('creates an error action if passed a non-boolean value', () => {
    expect(Actions.setCspEnabledAction('yes')).toMatchSnapshot()
  })
  it('creates a SET_CSP_ENABLED_OPTIMISTIC action when optimistic option is given', () => {
    expect(Actions.setCspEnabledAction(true, {optimistic: true})).toMatchSnapshot()
  })
})

describe('setCspEnabled', () => {
  it('converts non-plural contexts to plural', () => {
    const thunk = Actions.setCspEnabled('course', 1, true)
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      put: jest.fn(() => ({then() {}}))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeAxios.put).toHaveBeenCalledWith(expect.stringContaining('courses'), {
      status: 'enabled'
    })
  })

  it('does not modify plural contexts', () => {
    const thunk = Actions.setCspEnabled('courses', 1, true)
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      put: jest.fn(() => ({then() {}}))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeAxios.put).toHaveBeenCalledWith(expect.stringContaining('courses'), {
      status: 'enabled'
    })
  })

  it('dispatches an optimistic action followed by the final result', () => {
    const thunk = Actions.setCspEnabled('courses', 1, true)
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      put: jest.fn(() => ({
        then(func) {
          const fakeResponse = {data: {enabled: true}}
          func(fakeResponse)
        }
      }))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeDispatch).toHaveBeenNthCalledWith(1, {
      payload: true,
      type: 'SET_CSP_ENABLED_OPTIMISTIC'
    })
    expect(fakeDispatch).toHaveBeenNthCalledWith(2, {
      payload: true,
      type: 'SET_CSP_ENABLED'
    })
  })
})

describe('getCspEnabled', () => {
  it('converts non-plural contexts to plural', () => {
    const thunk = Actions.getCspEnabled('course', 1)
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      get: jest.fn(() => ({then() {}}))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeAxios.get).toHaveBeenCalledWith(expect.stringContaining('courses'))
  })

  it('dispatches a SET_CSP_ENABLED action when complete', () => {
    const thunk = Actions.getCspEnabled('courses', 1)
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      get: jest.fn(() => ({
        then(func) {
          const fakeResponse = {data: {enabled: true}}
          func(fakeResponse)
        }
      }))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeDispatch).toHaveBeenCalledWith({
      payload: true,
      type: 'SET_CSP_ENABLED'
    })
  })
})

describe('addDomainAction', () => {
  it('creates an ADD_DOMAIN action when passed string value', () => {
    expect(Actions.addDomainAction('instructure.com', 'account')).toMatchSnapshot()
  })
  it('creates an error action if passed a non-string value', () => {
    expect(Actions.addDomainAction(true, 'account')).toMatchSnapshot()
  })
  it('creates an error action if given an invalid domainType', () => {
    expect(Actions.addDomainAction('instructure', 'subaccount')).toMatchSnapshot()
  })

  it('creates an ADD_DOMAIN_OPTIMISTIC action when optimistic option is given', () => {
    expect(
      Actions.addDomainAction('instructure.com', 'account', {optimistic: true})
    ).toMatchSnapshot()
  })
})

describe('addDomain', () => {
  it('dispatches an optimistic action followed by the final result', () => {
    const thunk = Actions.addDomain('account', 1, 'instructure.com')
    const fakeDispatch = jest.fn()
    const fakeAxios = {
      post: jest.fn(() => ({
        then(func) {
          const fakeResponse = {}
          func(fakeResponse)
        }
      }))
    }
    thunk(fakeDispatch, null, {axios: fakeAxios})
    expect(fakeDispatch).toHaveBeenNthCalledWith(1, {
      payload: {account: 'instructure.com'},
      type: 'ADD_DOMAIN_OPTIMISTIC'
    })
    expect(fakeDispatch).toHaveBeenNthCalledWith(2, {
      payload: {account: 'instructure.com'},
      type: 'ADD_DOMAIN'
    })
  })
})

describe('addDomainBulkAction', () => {
  it('creates a ADD_DOMAIN_BULK action when passed an value', () => {
    expect(
      Actions.addDomainBulkAction({
        account: ['instructure.com'],
        tools: ['google.com']
      })
    ).toMatchSnapshot()
  })
  it('creates an error action if passed an invalid domainMap', () => {
    expect(
      Actions.addDomainBulkAction({
        lti: ['google.com']
      })
    ).toMatchSnapshot()
  })
})

describe('getCurrentWhitelist', () => {
  it('dispatches a bulk domain action ', () => {
    const thunk = Actions.getCurrentWhitelist('account', 1)
    const fakeDispatch = jest.fn()
    const fakeGetState = () => ({enabled: true})
    const fakeAxios = {
      get: jest.fn(() => ({
        then(func) {
          const fakeResponse = {
            data: {
              enabled: true,
              current_account_whitelist: ['instructure.com', 'canvaslms.com'],
              effective_whitelist: ['bridgelms.com'],
              tools_whitelist: ['lti-tool.com']
            }
          }
          func(fakeResponse)
        }
      }))
    }
    thunk(fakeDispatch, fakeGetState, {axios: fakeAxios})
    expect(fakeDispatch).toHaveBeenCalledWith({
      payload: {
        effective: ['bridgelms.com'],
        account: ['instructure.com', 'canvaslms.com'],
        tools: ['lti-tool.com']
      },
      type: 'ADD_DOMAIN_BULK'
    })
  })
})
