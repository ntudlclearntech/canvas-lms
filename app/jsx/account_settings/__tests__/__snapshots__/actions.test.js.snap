// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addDomainAction creates an ADD_DOMAIN action when passed string value 1`] = `
Object {
  "payload": Object {
    "account": "instructure.com",
  },
  "type": "ADD_DOMAIN",
}
`;

exports[`addDomainAction creates an ADD_DOMAIN_OPTIMISTIC action when optimistic option is given 1`] = `
Object {
  "payload": Object {
    "account": "instructure.com",
  },
  "type": "ADD_DOMAIN_OPTIMISTIC",
}
`;

exports[`addDomainAction creates an error action if given an invalid domainType 1`] = `
Object {
  "error": true,
  "payload": [Error: domainType is invalid],
  "type": "ADD_DOMAIN",
}
`;

exports[`addDomainAction creates an error action if passed a non-string value 1`] = `
Object {
  "error": true,
  "payload": [Error: Can only set to String values],
  "type": "ADD_DOMAIN",
}
`;

exports[`addDomainBulkAction creates a ADD_DOMAIN_BULK action when passed an value 1`] = `
Object {
  "payload": Object {
    "account": Array [
      "instructure.com",
    ],
    "tools": Array [
      "google.com",
    ],
  },
  "type": "ADD_DOMAIN_BULK",
}
`;

exports[`addDomainBulkAction creates an error action if passed an invalid domainMap 1`] = `
Object {
  "error": true,
  "payload": [Error: Invalid domain type key provided in domainsMap],
  "type": "ADD_DOMAIN_BULK",
}
`;

exports[`setCspEnabledAction creates a SET_CSP_ENABLED action when passed a boolean value 1`] = `
Object {
  "payload": true,
  "type": "SET_CSP_ENABLED",
}
`;

exports[`setCspEnabledAction creates a SET_CSP_ENABLED_OPTIMISTIC action when optimistic option is given 1`] = `
Object {
  "payload": true,
  "type": "SET_CSP_ENABLED_OPTIMISTIC",
}
`;

exports[`setCspEnabledAction creates an error action if passed a non-boolean value 1`] = `
Object {
  "error": true,
  "payload": [Error: Can only set to Boolean values],
  "type": "SET_CSP_ENABLED",
}
`;
